lib/streamlit/runtime/session_manager.py

```diff
@@ -0,0 +1,344 @@
+# consolidating it with the SessionData class since the two names are similar enough
+# to be confusing right now.
+@dataclass
+class SessionInfo:
+    """Type containing data related to an AppSession.
+
+    For each AppSession, the Runtime tracks that session's
+    script_run_count. This is used to track the age of messages in
+    the ForwardMsgCache.
+    """
+
+    client: Optional[SessionClient]
+    session: AppSession
+    script_run_count: int = 0
+
+
+class SessionStorageError(Exception):
+    """Exception class for errors raised by SessionStorage.
+
+    The original error that causes a SessionStorageError to be (re)raised will generally
+    be an I/O error specific to the concrete SessionStorage implementation.
+    """
+
+
+class SessionStorage(Protocol):
+    @abstractmethod
+    def get(self, session_id: str) -> Optional[SessionInfo]:
+        """Return the SessionInfo corresponding to session_id, or None if one does not
+        exist.
+
+        Parameters
+        ----------
+        session_id
+            The unique ID of the session being fetched.
+
+        Returns
+        -------
+        SessionInfo or None
+
+        Raises
+        ------
+        SessionStorageError
+            Raised if an error occurs while attempting to fetch the session.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def save(self, session_info: SessionInfo) -> None:
+        """Save the given session.
+
+        Parameters
+        ----------
+        session_info
+            The SessionInfo being saved.
+
+        Raises
+        ------
+        SessionStorageError
+            Raised if an error occurs while saving the given session.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def delete(self, session_id: str) -> None:
+        """Mark the session corresponding to session_id for deletion and stop tracking
+        it.
+
+        Note that:
+          * Calling delete on an ID corresponding to a nonexistent session is a no-op.
+          * Calling delete on an ID should cause the given session to no longer be
+            tracked by this SessionStorage, but exactly when and how the session's data
+            is eventually cleaned up is a detail left up to the implementation.
+
+        Parameters
+        ----------
+        session_id
+            The unique ID of the session to delete.
+
+        Raises
+        ------
+        SessionStorageError
+            Raised if an error occurs while attempting to delete the session.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def list(self) -> List[SessionInfo]:
+        """List all sessions tracked by this SessionStorage.
+
+        Returns
+        -------
+        List[SessionInfo]
+
+        Raises
+        ------
+        SessionStorageError
+            Raised if an error occurs while attempting to list sessions.
+        """
+        raise NotImplementedError
+
+
+class SessionManager(Protocol):
```

Comment:
It seems weird for a Protocol - which is a duck-typing mechanism (you don't need to explicitly inherit from a Protocol in order to conform to it) - to have concrete function implementations. I think this should instead just be an abstract base class (i.e. lose the `Protocol` designator).

Reply:
Hm, from the PEP defining protocol behavior (https://peps.python.org/pep-0544), it seems like it's reasonable to support both cases where protocols are given concrete implementations purely structurally and via explicit inheritance. There's a good amount of example code in the PEP defining protocols that have methods with default implementations, so doing this seems like it's relatively idiomatic.

Protocols at runtime become regular abstract base classes, and there's this relevant line from the PEP:
> The general philosophy is that protocols are mostly like regular ABCs, but a static type checker will handle them specially.

TL; DR - I think doing this is actually relatively Pythonic, so I'll probably keep things as-is unless you have a particularly strong opinion on this